common/lib/xmodule/xmodule/modulestore/split_mongo/split.py

```diff
@@ -2416,14 +2416,36 @@ def delete_item(self, usage_locator, user_id, force=False):
 
             return result
 
-    @contract(block_key=BlockKey, blocks='dict(BlockKey: BlockData)')
-    def _remove_subtree(self, block_key, blocks):
-        """
-        Remove the subtree rooted at block_key
-        """
-        for child in blocks[block_key].fields.get('children', []):
-            self._remove_subtree(BlockKey(*child), blocks)
-        del blocks[block_key]
+    @contract(root_block_key=BlockKey, blocks='dict(BlockKey: BlockData)')
+    def _remove_subtree(self, root_block_key, blocks):
+        """
+        Remove the subtree rooted at root_block_key
+        We do this breadth-first to make sure that we don't remove
+        any children that may have parents that we don't want to delete.
+        """
+        # create mapping from each child's key to its parents' keys
+        child_parent_map = defaultdict(list)
+        for block_key, block_data in blocks.iteritems():
+            for child in block_data.fields.get('children', []):
+                child_parent_map[BlockKey(*child)].append(block_key)
+
+        to_delete = {root_block_key}
+        tier = {root_block_key}
+        while tier:
+            next_tier = set()
+            for block_key in tier:
+                for child in blocks[block_key].fields.get('children', []):
+                    child_block_key = BlockKey(*child)
+                    parents = child_parent_map[child_block_key]
+                    # Make sure we want to delete all of the child's parents
+                    # before slating it for deletion
+                    if all(parent in to_delete for parent in parents):
+                        next_tier.add(child_block_key)
```

Comment:
Nit: if you made `child_parent_map` store sets, then you could just use `parents.issubset(to_delete)`.